Pull request-add biscuit aligner to the optional aligners, and relative steps

ekushele opened the pull request:
Many thanks to contributing to nf-core/methylseq!

Please fill in the appropriate checklist below (delete whatever is not relevant). These are the most common things requested on pull requests (PRs).

## PR checklist
 - [x] This comment contains a description of changes (with reason)
 - [x] Ensure the test suite passes (`nextflow run . -profile test,docker`).
 - [ ] Make sure your code lints (`nf-core lint .`) - had a bit of problem in the `.github/workflows/ci.yml`.
 - [x] Documentation in `docs` is updated
 - [x] `CHANGELOG.md` is updated
 - [x] `README.md` is updated 

**Learn more about contributing:** https://github.com/nf-core/methylseq/tree/master/.github/CONTRIBUTING.md

phue commented:
Hi @ekushele thanks alot for adding this!

My major concern with this is that it will break the conda profile.
I think the cleanest way to solve this, to contribute a [bioconda recipe](https://bioconda.github.io/contributor/index.html) for biscuit.
What do you think?

ekushele commented:
Thanks @phue !
I created bioconda recipe file, and I made a PR to the github bioconda.
Now I'm waiting for them to review it..
I'll update the docker and environment file when it'll be available.

Thanks again!

ewels commented:
x-ref https://github.com/bioconda/bioconda-recipes/pull/21615

ekushele commented:
How can I update my PR to include the new commit with the new Dockerfile and environment.yml file?
Or it was updated automatically?

phue commented:
Great! Thanks for adding the bioconda recipe!

> How can I update my PR to include the new commit with the new Dockerfile and environment.yml file?
> Or it was updated automatically?

Yes, this PR does get updated automatically everytime you push to your branch.
Before we start reviewing this, I'm afraid there are some conflicts that need to be resolved. Can you try merging the current `nf-core/methylseq:dev` branch into yours? Also it would be helpful to add the `biscuit` aligner option to the CI github action so we can have tests running here

ekushele commented:
So I merged the "dev" branch into mine, and solved the conflicts.
I hope I did well when adding the biscuit to the CI github action: added it to the list of aligners, and  added `--bwa_biscuit_index  results/reference_genome/genome.fa` which I understood to be the result output that should come out. Am I right on that?

phue commented on bin/setup.sh:
```diff
@@ -0,0 +1,39 @@
+#!/usr/bin/env bash
```
Where is this file sourced? Do we really need it? If the variables are needed, nextflow's [env scope](https://www.nextflow.io/docs/latest/config.html#scope-env) should be considered

ekushele replied:
I don't really understand the question...
The code was taken from the biscuit-scripts section, in [here](https://github.com/huishenlab/biscuit/tree/master/scripts)
This file is part of the biscuit_QC process, and needed for the QC.sh file. The variable are set through the QC.sh script, so I don't really understand the second part...

phue replied:
I meant that I can't find the place where `setup.sh` is executed. All it seems to do is setting some environment variables, which are not even used in `QC.sh`

ekushele replied:
They are used and referred to in QC.sh.
Can one unit the files? maybe yes, it's only setting of variables for the QC.sh

phue replied:
Looking at https://github.com/huishenlab/biscuit/blob/master/scripts/QC.sh
I don't see a reference to any of these variables

ekushele replied:
Yes. The biscuit team told me it's "a new QC script which doesn't need the setup script. You are welcome to use it, but it is incompatible with the current version of MultiQC", so until they update the MultiQC, they suggested me to keep using my own script, which is the older verstion.
Anyway, I'm working on adding it to the recipe, so it will be update automatically... No?

ewels replied:
It would be even better to add support to MultiQC instead of writing your own script, if possible. Then everyone benefits 😉 

I'm doing a period of work on MultiQC at the moment and reviewing pull requests, will hopefully be getting a new release out relatively soon.

ekushele replied:
It's not my own script. It's the biscuit QC.sh old script, which now has support in the MultiQC.
The biscuit team changed the QC.sh script, so the newer script, and they need to update also the MultiQC and they didn't do it yet.

phue commented on docs/output.md:
```diff
@@ -57,7 +60,7 @@ Single-end data will have slightly different file names and only one FastQ file
 ### Alignment
 
-Bismark and bwa-meth convert all Cytosines contained within the sequenced reads to Thymine _in-silico_ and then align against a three-letter reference genome. This method avoids methylation-specific alignment bias. The alignment produces a BAM file of genomic alignments.
+Bismark and bwa-meth convert all Cytosines contained within the sequenced reads to Thymine _in-silico_ and then align against a three-letter reference genome. This method avoids methylation-specific alignment bias. The alignment produces a BAM file of genomic alignments. _+__________
```
```suggestion
Bismark and bwa-meth convert all Cytosines contained within the sequenced reads to Thymine _in-silico_ and then align against a three-letter reference genome. This method avoids methylation-specific alignment bias. The alignment produces a BAM file of genomic alignments. 
```

phue commented on docs/usage.md:
```diff
@@ -74,14 +81,16 @@ It is recommended to limit the Nextflow Java virtual machines memory. We recomme
+### Bismark, bwa-meth and biscuit workflow
 
-The nf-core/methylseq package is actually two pipelines in one. The default workflow uses [Bismark](http://www.bioinformatics.babraham.ac.uk/projects/bismark/) with [Bowtie2](http://bowtie-bio.sourceforge.net/bowtie2/index.shtml) as alignment tool: unless specified otherwise, nf-core/methylseq will run this pipeline.
+The nf-core/methylseq package is actually threepipelines in one. The default workflow uses [Bismark](http://www.bioinformatics.babraham.ac.uk/projects/bismark/) with [Bowtie2](http://bowtie-bio.sourceforge.net/bowtie2/index.shtml) as alignment tool: unless specified otherwise, nf-core/methylseq will run this pipeline.
```
```suggestion
The nf-core/methylseq package is actually three pipelines in one. The default workflow uses [Bismark](http://www.bioinformatics.babraham.ac.uk/projects/bismark/) with [Bowtie2](http://bowtie-bio.sourceforge.net/bowtie2/index.shtml) as alignment tool: unless specified otherwise, nf-core/methylseq will run this pipeline.
```

phue commented on docs/usage.md:
```diff
@@ -31,6 +33,11 @@
   * [`--unmapped`](#--unmapped)
   * [`--save_trimmed`](#--save_trimmed)
   * [`--save_align_intermeds`](#--save_align_intermeds)
+  * [`--save_pileup_file`](#--save_pileup_file)
+  * [`--epiread`](#--epiread)
+  * [`--save_snp_file`](#--save_snp_file)
+  * [`--soloWCGW_file`](#--soloWCGW_file)
+  * [`--assets_dir`](#--assets_dir)
```
these parameters should also have standard values

ekushele replied:
What does it mean? sorry for not understanding...

phue replied:
nextflow will throw ugly warnings if a parameter is not initialised to a default value (such as `false`)

ekushele replied:
Oh OK, so I can initialized all this parameters to true. or false...

phue commented on docs/usage.md:
```diff
@@ -325,9 +344,33 @@ By default, trimmed FastQ files will not be saved to the results directory. Spec
+
+### `--soloWCGW_file`
+
+This will generatea  methylation statuses in [bedGraph](http://genome.ucsc.edu/goldenPath/help/bedgraph.html) format, intersected with soloWCGW after extracting methylation from vcf, using biscuit workflow.
```
```suggestion
This will generate methylation statuses in [bedGraph](http://genome.ucsc.edu/goldenPath/help/bedgraph.html) format, intersected with soloWCGW after extracting methylation from vcf, using biscuit workflow.
```

phue commented on docs/usage.md:
```diff
@@ -325,9 +344,33 @@ By default, trimmed FastQ files will not be saved to the results directory. Spec
+
+### `--assets_dir`
+
+Path to a directory containing needed file for biscuit-QC step. The needed files for hg38,hg19 and mm10 can be found in [here](https://www.cse.huji.ac.il/~ekushele/assets.html). 
```
How where these files generated?

phue commented on docs/usage.md:
```diff
@@ -325,9 +344,33 @@ By default, trimmed FastQ files will not be saved to the results directory. Spec
+### `--assets_dir`
+
+Path to a directory containing needed file for biscuit-QC step. The needed files for hg38,hg19 and mm10 can be found in [here](https://www.cse.huji.ac.il/~ekushele/assets.html). 
+**This paramater is mandatory when running the pipeline using biscuit workflow**
```
```suggestion
**This parameter is mandatory when running the pipeline using biscuit workflow**
```

phue commented on main.nf:
```diff
@@ -30,28 +30,32 @@ def helpMessage() {
      --local_alignment [bool]           Allow soft-clipping of reads (potentially useful for single-cell experiments)
-     --bismark_align_cpu_per_multicore [int] Specify how many CPUs are required per --multicore for bismark align (default = 3)
-     --bismark_align_mem_per_multicore [str] Specify how much memory is required per --multicore for bismark align (default = 13.GB)
+	 --soloWCGW_file [path]             soloWCGW file, to intersect with methyl_extract bed file. soloWCGW for hg38 can be downlaod from: www.cse.huji.ac.il/~ekushele/solo_WCGW_cpg_hg38.bed. EXPERMINTAL!
```
How was this file generated?

ekushele replied:
I removed this step, because it's a bit complicated to explain how they were generated... Done

phue commented on main.nf:
```diff
@@ -486,27 +581,18 @@ if( params.skip_trimming ){
         file "where_are_my_files.txt"
 
         script:
-        def c_r1 = clip_r1 > 0 ? "--clip_r1 $clip_r1" : ''
```
why did you remove the multicore support?

ekushele replied:
I think I just forgot to copy the newer version of methylseq to this process. Fixed

phue commented on main.nf:
```diff
@@ -675,7 +815,7 @@ if( params.aligner =~ /bismark/ ){
         if(params.single_end) {
```
Why did you remove the cytosine report parameter?

ekushele replied:
Again, didn't see it exists in the newer versions. I didn't touch this step myself... Fixed

phue commented on main.nf:
```diff
@@ -914,17 +1055,309 @@ else {
+		script:
+		assembly = fasta.toString().replaceAll(/\.\w+/,"")
+		"""
+		$baseDir/bin/biscuit_QC.sh -v ${vcf[0]} -o ${bam_name}.${assembly}_biscuitQC $assets $fasta ${bam_name}.${assembly} ${bam} -p ${task.cpus}
```
I assume this script is taken from [here](https://github.com/zhou-lab/biscuit/tree/master/scripts)?

What happens in case of updates? We would need to manually copy the upstream script again.
I don't like that, it would be better to call it from the conda env for instance

ekushele replied:
make sense. How can one do that?

phue replied:
It would need to be included in the conda recipe and added to $PATH

phue commented on main.nf:
```diff
@@ -933,23 +1366,115 @@ process qualimap {
+
+      input:
+      file fasta from ch_fasta_for_picard
+	file file_try from ch_try
```
what is this used for?

ekushele replied:
I wanted to make sure that "get_software_versions" run first in the list of processes, because I had some problems with that... Sorry, my mistake I forgot to remove that. Fixed

phue commented on CHANGELOG.md:
```diff
@@ -1,5 +1,20 @@
+
+* Added Picard CollectInsertSizeMetrics and Picard CollectGcBiasMetrics
+* Improved qulimap and preseq by adding `samtools sort` and `samtools index` step in the Bismark aligner
+* Added biscuit aligner as an optional aligner, with all relative steps (alignment, mark duplicates, methylation extraction, QC for biscuit, and opptional epiread file creation).
```
```suggestion
* Added biscuit aligner as an optional aligner, with all relative steps (alignment, mark duplicates, methylation extraction, QC for biscuit, and optional epiread file creation).
```

phue commented on CHANGELOG.md:
```diff
@@ -1,5 +1,20 @@
+### Software updates
+
+* _new dependency_: samblaster`0.1.24`
+* _new dependect_: bedtools `2.29.1`
```
```suggestion
* _new dependency_: bedtools `2.29.1`
```

phue commented on docs/output.md:
```diff
@@ -101,6 +122,7 @@ This step removes alignments with identical mapping position to avoid technical
 **bwa-meth output directory: `results/bwa-mem_markDuplicates/`**
 
 > **NB:** The bwa-meth step doesn't remove duplicate reads from the BAM file, it just labels them.
+> 
```
```suggestion
```

phue commented on docs/output.md:
```diff
@@ -109,11 +131,21 @@ This step removes alignments with identical mapping position to avoid technical
 The methylation extractor step takes a BAM file with aligned reads and generates files containing cytosine methylation calls. It produces a few different output formats, described below.
 
-Note that the output may vary a little depending on whether you specify `--comprehensive` or `--non_directional` when running the pipeline.
+Note that the output may vary a little depending on whether you specify `--comprehensive` or `--non_directional`  or `--skip_deduplication` or `--rrbs`when running the pipeline.
```
```suggestion
Note that the output may vary a little depending on whether you specify `--comprehensive` or `--non_directional`  or `--skip_deduplication` or `--rrbs` when running the pipeline.
```

phue commented on docs/output.md:
```diff
@@ -142,6 +174,28 @@ Filename abbreviations stand for the following reference alignment strands:
+	> * `sample.epiread` 
+			 Storing CpG retention pattern as well as SNP information on the same read in a compact way 
+> * `snp/sample.snp.bed`
+	> 	Storing the SNP information of the `sample.epiread` file. **EXTRACTING SNP PROCSS IS IN PROGRESS!**
```
Is this something you want to add to this PR?

ekushele replied:
Hmm.. I'll remove that until the biscuit team finish to work on it...

phue commented on docs/usage.md:
```diff
@@ -215,7 +224,7 @@ params {
 If you don't want to use the Illumina iGenomes references, you can supply your own reference genome.
 
-The minimum requirement is just a FASTA file - the pipeline will automatically generate the relevant reference index from this. You can use the command line option `--save_reference` to keep the generated references so that they can be added to your config and used again in the future. The bwa-meth workflow always needs a FASTA file, for methylation calling.
+The minimum requirement is just a FASTA file - the pipeline will automatically generate the relevant reference index from this. You can use the command line option `--save_reference` to keep the generated references so that they can be added to your config and used again in the future. The bwa-meth and biscuit workflows always need a FASTA file, for methylation calling. The FASTA is also required ...
```
```suggestion
The minimum requirement is just a FASTA file - the pipeline will automatically generate the relevant reference index from this. You can use the command line option `--save_reference` to keep the generated references so that they can be added to your config and used again in the future. The bwa-meth and biscuit workflows always need a FASTA file, for methylation calling. The FASTA is also required for the generation of Picard metrics.
```
